Conjur/conventions/ruby-style-guide.md

```diff
@@ -24,10 +23,194 @@ This guide is intended to be a "living document" in the sense that it's expected
+
+  - Bad:
+
+    ```ruby
+    def get_jwt_identity_from_request
+        @logger.debug(LogMessages::Authentication::AuthnJwt::CallingGetJwtIdentity.new)
+        @jwt_identity ||= @jwt_configuration.jwt_identity
+        @logger.info(LogMessages::Authentication::AuthnJwt::FoundJwtIdentity.new(jwt_identity))
+    end
+    ```
+
+  - Good:
+
+    ```ruby
+    def get_jwt_identity_from_request
+        @logger.debug(LogMessages::Authentication::AuthnJwt::CallingGetJwtIdentity.new)
+        jwt_identity
+        @logger.info(LogMessages::Authentication::AuthnJwt::FoundJwtIdentity.new(jwt_identity))
+    end
+    
+    def jwt_identity
+        @jwt_identity ||= @jwt_configuration.jwt_identity
+    end
+    ```
+
+- In constuctor of classes separate input handling and dependency handling with empty line
+
+- In command class implement methods in the order they are called
+
+  ```ruby
+  def call
+    func_a
+    func_b
+    func_c
+  end
+  
+  private
+  
+  def func_a
+    ...
+  end
+  
+  def func_b
+    ...
+  end
+  
+  def sub_func_b
+    ...
+  end
+  
+  def func_c
+    ...
+    end
+  ```
+
+- In functions with many parameters put each parameter in seperate line in the declaration. In this case end each param with ":" so the caller will need to specify name of variable for each param.
+
+  ```
+  def method(
+    param1:
+    param2:
+    param3:
+  )
+  ```
+
+  
+
+- When overriding a code climate violation with an inline exception (this isn't possible for all code climate  plugins but is for many), add the reason for the exception on the line above the special "ignore" directive
+
+   https://github.com/cyberark/conjur/blob/5fe6b22f225a1f2cf0be90d90f184c74c150be5a/cucumber/policy/features/support/rest_client_wrapper.rb#L12
+
+## Design Baseline
+
+- On any new class ask yourself if it should be a class or command class. Command Classes coordinate the execution of a multi-step process.
+
+  https://github.com/cyberark/conjur/blob/master/app/domain/README.md#the-functional-object-oriented-style
+
+  
+
+- No need for interface classes
+
+- *Never* use inheritance or mixins.
+
+- Dependencies for classes should be passed as default values in constructor
+
+  ```ruby
+  def initialize(
+    authenticator_input:,
+    logger: Rails.logger,
+    validate_resource_restrictions_class: Authentication::ResourceRestrictions::ValidateResourceRestrictions,
+    extract_token_from_credentials: Authentication::AuthnJwt::InputValidation::ExtractTokenFromCredentials.new,
+    create_identity_provider: Authentication::AuthnJwt::IdentityProviders::CreateIdentityProvider.new
+  )
+  ```
+
+- Use memoization when you need to run "hard operation" more than one and you sure that the output is the same (DB queries, API requests, etc..) . Use the '||=' to do it. The name of the function in memoization and the name of the memoaized object should be the same.
```

Comment:
```suggestion
- Use memoization when you need to run "expensive operation" more than once and the output is the same (DB queries, API requests, etc..) . Use the '||=' to do it. The name of the function in memoization and the name of the memoized object should be the same.
```

Mostly interested in swapping "hard" for "expensive". Memoization increases memory consumption but is a small price to pay to avoid the hit of re-querying an API for the same data.